### PR TITLE
Replace certificate chain info query usage with certificate download

### PR DESCRIPTION
**Pull request by andresilva91**
## Motivation

Now that we can download certificates, using the chain info query isn't needed anymore

## Proposal

Use the download certificate entrypoint instead

## Test Plan

CI

### CODE REVIEW

**Comment by andresilva91** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/linera-io/linera-protocol/2081?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#2081** <a href="https://app.graphite.dev/github/pr/linera-io/linera-protocol/2081?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#2079** <a href="https://app.graphite.dev/github/pr/linera-io/linera-protocol/2079?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2074** <a href="https://app.graphite.dev/github/pr/linera-io/linera-protocol/2074?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @andresilva91 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: linera-core/src/local_node.rs [hunk 1]
```diff
@@ -75,6 +76,11 @@ pub enum LocalNodeError {
 
     #[error("Got different blob id from downloaded blob when trying to download {blob_id:?}")]
     BlobIdMismatch { blob_id: BlobId },
+
+    #[error(
+        "Got different hash from downloaded certificate value when trying to download {hash:?}"
+    )]
+    CertificateValueHashMismatch { hash: CryptoHash },
```
Comment (MathieuDutSik) [inline linera-core/src/local_node.rs:83]: If they are different, maybe one could put both values in the error?
That could help the debugging.

File: linera-core/src/local_node.rs [hunk 2]
```diff
@@ -634,25 +625,22 @@ where
...
-        chain_id: ChainId,
+        mut validators: Vec<A>,
         location: BytecodeLocation,
-    ) -> Option<HashedCertificateValue>
+    ) -> Result<Option<HashedCertificateValue>, LocalNodeError>
     where
         A: LocalValidatorNode + Clone + 'static,
     {
         // Sequentially try each validator in random order.
         validators.shuffle(&mut rand::thread_rng());
```
Comment (MathieuDutSik) [inline linera-core/src/local_node.rs:635]: Maybe explain why applying the random shuffle?

File: linera-core/src/client.rs [hunk 3]
```diff
@@ -1307,12 +1305,17 @@ where
...
+            .validator_node_provider
+            .make_nodes::<Vec<_>>(&committee)?;
         let values = self
             .client
             .local_node
-            .read_or_download_hashed_certificate_values(nodes.clone(), block.bytecode_locations())
+            .clone()
+            .read_or_download_hashed_certificate_values(
+                nodes.clone(),
+                block.bytecode_locations().into_keys(),
```
Comment (afck) [inline linera-core/src/client.rs:1317]: If we don't need the sender chain ID anymore, should we return only the locations from `Block::bytecode_locations`?

File: linera-core/src/local_node.rs [hunk 4]
```diff
@@ -594,7 +584,7 @@ where
     }
 
     pub async fn try_synchronize_chain_state_from<A>(
-        &self,
+        &mut self,
```
Comment (afck) [inline linera-core/src/local_node.rs:587]: Does this still need to be mutable?

Comment (andresilva91) [reply linera-core/src/local_node.rs:587]: Will remove!

File: linera-core/src/local_node.rs [hunk 5]
```diff
@@ -720,21 +707,33 @@ where
     }
 
     async fn try_download_hashed_certificate_value_from<A>(
-        name: ValidatorName,
         node: &mut A,
-        chain_id: ChainId,
+        name: ValidatorName,
         location: BytecodeLocation,
-    ) -> Option<HashedCertificateValue>
+    ) -> Result<Option<HashedCertificateValue>, LocalNodeError>
```
[the remark this answers is not in the document]
Comment (andresilva91) [reply linera-core/src/local_node.rs:713]: Good point! Will remove also

File: linera-core/src/client.rs (hunk 3 above)
Comment (andresilva91) [reply linera-core/src/client.rs:1317]: Good point!

File: linera-core/src/local_node.rs [hunk 6]
```diff
@@ -720,21 +706,33 @@ where
...
-            ChainInfoQuery::new(chain_id).with_hashed_certificate_value(location.certificate_hash);
-        if let Ok(response) = node.handle_chain_info_query(query).await {
-            if response.check(name).is_ok() {
-                return response.info.requested_hashed_certificate_value;
+        match node
+            .download_certificate_value(location.certificate_hash)
+            .await
+        {
+            Ok(hashed_certificate_value) => {
+                if hashed_certificate_value.hash() == location.certificate_hash {
```
Comment (afck) [inline linera-core/src/local_node.rs:721]: Is this redundant now?

**Review by afck — APPROVED** [review]

File: linera-core/src/local_node.rs (hunk 6 above)
Comment (andresilva91) [reply linera-core/src/local_node.rs:721]: It is!

**Comment by andresilva91** [comment]
### Merge activity

* **Jun 18, 8:14 AM EDT**: @andresilva91 started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/linera-io/linera-protocol/2081).

**Pull request merged**